File: leetcode/permutationsII.py
# Given a collection of numbers that might contain duplicates, 
# return all possible unique permutations.
# Time Complexity: O((n+1)!)
# Space Complexity: O((n+1)!)

class Solution:
    # @param {integer[]} nums
    # @return {integer[][]}
    def permuteUnique(self, nums):
        nums.sort()
        result = [list(nums)]
        while True:
            i = len(nums) - 2
            while i >= 0 and nums[i] >= nums[i + 1]:
                i -= 1
            if i < 0: break
            j = len(nums) - 1
            while nums[i] >= nums[j]: 
                j -= 1
            nums[i], nums[j] = nums[j], nums[i]
            left, right = i + 1, len(nums) - 1
            nums[i + 1:] = nums[-1:i:-1]
            result.append(list(nums))
        return result

if __name__ == "__main__":
    solution = Solution()
    print solution.permuteUnique([1, 1, 2])

File: leetcode/permutationsII.js
/* Given a collection of numbers that might contain duplicates, 
 * return all possible unique permutations.
 * Time Complexity: O((n+1)!)
 * Space Complexity: O((n+1)!)
 */

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    nums.sort();
    for (var result = [nums.slice()]; true; ) {
        for (var i = nums.length - 2; i >= 0 && nums[i] >= nums[i + 1]; --i)
            ;
        if (i < 0)
            break;
        for (var j = nums.length - 1; nums[i] >= nums[j]; --j)
            ;
        var temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
        nums = nums.slice(0, i + 1).concat(nums.slice(i + 1).reverse());
        result.push(nums.slice());
    }
    return result;
};
